include/xsimd/types/xsimd_avx_shuffling.hpp

```diff
@@ -0,0 +1,148 @@
+/***************************************************************************
+* Copyright (c) 2016, Johan Mabille and Sylvain Corlay                     *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
+#ifndef XSIMD_AVX_SHUFFLING_HPP
+#define XSIMD_AVX_SHUFFLING_HPP
+
+#include "xsimd/types/xsimd_avx_conversion.hpp"
+
+namespace xsimd 
+{
+    // Version of select(a, b) with compile time mask
+    template <class B>
+    struct selector_immediate;
+
+    template <class T, std::size_t N>
+    struct selector_immediate<batch<T, N>>
+    {
+        using type = int32_t;
+    };
+
+    template <typename selector_immediate<batch<float, 8>>::type P, class T>
+    batch<T, 8> select(const batch<T, 8>& a, const batch<T, 8>& b);
+
+    template <typename selector_immediate<batch<double, 4>>::type P, class T>
+    batch<T, 4> select(const batch<T, 4>& a, const batch<T, 4>& b);
+
+    constexpr static int AaBbCcDd = 1;
+    constexpr static int AaCc = 2;
+    constexpr static int BbDd = 3;
+    constexpr static int AaBbEeFf = 4;
+    constexpr static int ABabEFef = 5;
+    constexpr static int CcDdGgHh = 6;
+    constexpr static int CDcdGHgh = 7;
+
+    template <uint64_t P, int B = -1, class T, size_t N>
+    auto interleave(const batch<T, N>& a, const batch<T, N>& b);
+
+    namespace detail
+    {
+        template <uint64_t P, int B, class T, size_t N>
+        struct interleave_impl;
+
+        template <class T, size_t N>
+        struct interleave_impl<AaCc, 64, T, N>
+        {
+            batch<T, N> operator()(const batch<T, N>& a, const batch<T, N>& b)
+            {
+                // {a, b, c, d} Δ {A, B, C, D} → {a, A, c, C}
+                batch<int64_t, 4> tmp = _mm256_unpacklo_epi64(bitwise_cast<batch<int64_t, 4>>(a),
+                                                              bitwise_cast<batch<int64_t, 4>>(b));
+                return bitwise_cast<batch<T, N>>(tmp);
```

Comment:
My reading of the unpacklo/unpackhi spec is that they actually do {a, A, b, B} and {c, C, d, D} respectively

Reply:
unpacklo/hi on mm256 work on two different mm128 vectors and do the same transformation on each. therefore unpacklo does aA on mm128 1 and cC in the second.

Reply:
Oh yes, you're right, it's actually even weirder than I thought. If I take _mm256_unpacklo_ps, for example, it is defined as...

```
INTERLEAVE_DWORDS(src1[127:0], src2[127:0]){
  dst[31:0] := src1[31:0]
  dst[63:32] := src2[31:0]
  dst[95:64] := src1[63:32]
  dst[127:96] := src2[63:32]
  RETURN dst[127:0]
}

dst[127:0] := INTERLEAVE_DWORDS(a[127:0], b[127:0])
dst[255:128] := INTERLEAVE_DWORDS(a[255:128], b[255:128])
dst[MAX:256] := 0
```

Which means that if you take as input {a, b, c, d, e, f, g, h} and {A, B, C, D, E, F, G, H}, you will do...

```
dst[127:0] := INTERLEAVE_DWORDS({a, b, c, d}, {A, B, C, D}) == {a, A, b, B}
dst[255:128] := INTERLEAVE_DWORDS({e, f, g, h}, {E, F, G, H}) == {e, E, f, F}
```

So the end result is indeed ({a, A, b, B, e, E, f, F}).

Reply:
I think I'm starting to understand why the Boost.SIMD devs just threw the towel and said "okay, just give us an arbitrary compile-time permutation and we'll compile it down into whatever the host CPU supports" :)

Reply:
What would be cool is to write a program with some optimizer that can find the optimal permutation sequence -- it could be written in python and use the cost of the instruction docs from agner fog :) (In case you're looking for a research project, haha).

Reply:
There is some prior art on this: https://users.ece.cmu.edu/~franzf/papers/cc08.pdf

Reply:
I'm trying to move some code from Boost.SIMD to xsimd and so far this is the only kind of stuff that I a missing.
trying to shuffle<0, 2, 4, 6, 1, 3, 5, 7> a batch<int16_t, 8>

Would appreciate a split_low and split_high as well :)

Thanks !